testing/integration_testing/restore_secrets.py

```diff
@@ -0,0 +1,161 @@
+  integration_test directories.
+- The server key will be patched into the Messaging project.
+- The uri prefix will be patched into the Dynamic Links project.
+- The reverse id will be patched into all Info.plist files, using the value from
+  the decrypted Google Service plist files as the source of truth.
+
+"""
+
+import argparse
+import os
+import plistlib
+import subprocess
+
+parser = argparse.ArgumentParser(
+    description="Decrypt and restore the secrets. Must specify one of"
+    " passphrase or passphrase_file.")
+parser.add_argument("--passphrase", help="The passphrase itself.")
+parser.add_argument("--passphrase_file", help="Path to file with passphrase.")
+parser.add_argument("--repo_dir", default=os.getcwd(), help="Path to SDK Repo")
+args = parser.parse_args()
+
+
+def main():
+  repo_dir = args.repo_dir
+  # The passphrase is sensitive, do not log.
+  if args.passphrase:
+    passphrase = args.passphrase
+  elif args.passphrase_file:
+    with open(args.passphrase_file, "r") as f:
+      passphrase = f.read()
+  else:
+    raise ValueError("Must supply passphrase or passphrase_file arg.")
+
+  secrets_dir = os.path.join(repo_dir, "scripts", "gha-encrypted")
+  encrypted_files = _find_encrypted_files(secrets_dir)
+  print("Found these encrypted files:\n%s" % "\n".join(encrypted_files))
+
+  for path in encrypted_files:
+    if "google-services" in path or "GoogleService" in path:
+      print("Encrypted Google Service file found: %s" % path)
+      # We infer the destination from the file's directory, example:
+      # /scripts/gha-encrypted/auth/google-services.json.gpg turns into
+      # /<repo_dir>/auth/integration_test/google-services.json
+      api = os.path.basename(os.path.dirname(path))
+      file_name = os.path.basename(path).replace(".gpg", "")
+      dest_path = os.path.join(repo_dir, api, "integration_test", file_name)
+      decrypted_text = _decrypt(path, passphrase)
+      with open(dest_path, "w") as f:
+        f.write(decrypted_text)
+      print("Copied decrypted google service file to %s" % dest_path)
+      # We use a Google Service file as the source of truth for the reverse id
+      # that needs to be patched into the Info.plist files.
+      if dest_path.endswith(".plist"):
+        _patch_reverse_id(dest_path)
+
+  print("Attempting to patch Dynamic Links uri prefix.")
+  uri_path = os.path.join(secrets_dir, "dynamic_links", "uri_prefix.txt.gpg")
+  uri_prefix = _decrypt(uri_path, passphrase)
+  dlinks_project = os.path.join(repo_dir, "dynamic_links", "integration_test")
+  _patch_main_src(dlinks_project, "REPLACE_WITH_YOUR_URI_PREFIX", uri_prefix)
+
+  print("Attempting to patch Messaging server key.")
+  server_key_path = os.path.join(secrets_dir, "messaging", "server_key.txt.gpg")
+  server_key = _decrypt(server_key_path, passphrase)
+  messaging_project = os.path.join(repo_dir, "messaging", "integration_test")
+  _patch_main_src(messaging_project, "REPLACE_WITH_YOUR_SERVER_KEY", server_key)
```

Comment:
Hmm, I wonder if we should actually formalize and factor this out. Scan everything for "REPLACE_WITH_YOUR_* and replace with the string in the relevant text file, if present. WDYT? (just a thought, not a strong suggestion).

Reply:
I think there's some merit to that idea, though the possibility of matching something it's not supposed to and patching a secret into an unexpected location (possibly a README file mentioning the placeholder, for example) makes me nervous. Or running into some other unexpected issue that I'm not currently predicting, like running into a file it doesn't have permission to read.

Reply:
Yeah, fair. I would like to see this more data-driven though. But let's file that in the backlog.